src/js/actions/libraries.js

```diff
@@ -140,59 +174,66 @@ define(function (require, exports) {
-            var layerFileName = currentLayer.smartObject.fileReference,
-                fileExtension = pathUtil.extension(layerFileName);
+        if (currentLayers.size === 1 && firstLayer.isSmartObject()) {
+            var layerFileName = firstLayer.smartObject.fileReference;
+
+            // layerFileName will be undefined if CC Libraries is uploading the same layer.
+            if (!layerFileName) {
+                return Promise.resolve();
+            }
+
+            var fileExtension = pathUtil.extension(layerFileName);
 
             if (_EXTENSION_TO_REPRESENTATION_MAP.hasOwnProperty(fileExtension)) {
                 representationType = _EXTENSION_TO_REPRESENTATION_MAP[fileExtension];
             }
         }
 
-        var tempLayerName = (new Date().getTime()).toString();
-
-        return os.getTempFilename(tempLayerName)
-            .bind(this)
-            .then(function (tempFilename) {
+        return _getTempPaths().bind(this)
+            .then(function (paths) {
                 // Export the selected layers
 
-                var tempPath = path.dirname(tempFilename.path),
-                    tempPreviewPath = tempPath + "/preview.png",
-                    previewSize = { w: 248, h: 188 },
-                    exportObj = libraryAdapter.exportLayer(tempPath, tempPreviewPath, tempLayerName, previewSize);
+                var previewSize = { w: RENDITION_SIZE, h: RENDITION_SIZE },
+                    exportObj = libraryAdapter.exportLayer(paths.tempBasePath, paths.tempPreviewPath,
+                        paths.tempName, previewSize);
 
-                return descriptor.playObject(exportObj);
+                return descriptor.playObject(exportObj).then(function (saveData) {
+                    paths.exportedLayerPath = saveData.in._path;
+                    return paths;
+                });
             })
-            .then(function (saveData) {
-                // Create new graphic asset of the exported layer(s) using the CC Libraries api.
+            .then(function (paths) {
+                // Create new graphic asset of the exported layer(s) using the CC Libraries api. 
 
                 currentLibrary.beginOperation();
-                newElement = currentLibrary.createElement(currentLayer.name, IMAGE_ELEMENT_TYPE);
+                newElement = currentLibrary.createElement(firstLayer.name, IMAGE_ELEMENT_TYPE);
 
-                return Promise.fromNode(function (cb) {
-                    var exportedLayerPath = saveData.in._path,
-                        representation = newElement.createRepresentation(representationType, "primary");
+                return Promise.fromNode(function (done) {
+                    var representation = newElement.createRepresentation(representationType, "primary");
 
-                    representation.updateContentFromPath(exportedLayerPath, false, cb);
+                    representation.updateContentFromPath(paths.exportedLayerPath, false, function () {
```

Comment:
You got a node style callback mixed in with promises here. YOu can pass done to updateContentFromPath, and then call setRenditionCache in the next promise block.